File: src/Core/Process/Factor.fs
namespace ARCtrl.Process

open ARCtrl
open ARCtrl.Helper
open Fable.Core

[<AttachMembers>]
type Factor = 
    {
        Name : string option
        FactorType : OntologyAnnotation option
        Comments : ResizeArray<Comment> option
    }

    static member make name factorType comments =
        {
            Name    = name
            FactorType = factorType
            Comments = comments         
        }

    static member create(?Name,?FactorType,?Comments) : Factor =
        Factor.make Name FactorType Comments

    static member empty =
        Factor.create()

    /// Create a ISAJson Factor from ISATab string entries
    static member fromString (name : string, term:string, source:string, accession:string, ?comments : ResizeArray<Comment>) =
        let oa = OntologyAnnotation.create (term, source, accession, ?comments = comments)
        Factor.make (Option.fromValueWithDefault "" name) (Option.fromValueWithDefault (OntologyAnnotation()) oa) None

    /// Get ISATab string entries from an ISAJson Factor object
    static member toStringObject (factor : Factor) =
        factor.FactorType |> Option.map OntologyAnnotation.toStringObject |> Option.defaultValue {|TermName = ""; TermAccessionNumber = ""; TermSourceREF = ""|}

    member this.NameText =
        this.Name
        |> Option.defaultValue ""

    member this.MapCategory(f : OntologyAnnotation -> OntologyAnnotation) =
        {this with FactorType = Option.map f this.FactorType}

    member this.SetCategory(c : OntologyAnnotation) =
        {this with FactorType = Some c}

    interface IISAPrintable with
        member this.Print() =
            this.ToString()
        member this.PrintCompact() =
            "OA " + this.NameText

    /// If a factor with the given name exists in the list, returns it
    static member tryGetByName (name : string) (factors:Factor list) =
        List.tryFind (fun (f:Factor) -> f.Name = Some name) factors

    /// If a factor with the given name exists in the list exists, returns true
    static member existsByName (name : string) (factors:Factor list) =
        List.exists (fun (f:Factor) -> f.Name = Some name) factors

    /// adds the given factor to the factors  
    static member add (factors:Factor list) (factor : Factor) =
        List.append factors [factor]

    /// If a factor with the given name exists in the list, removes it
    static member removeByName (name : string) (factors : Factor list) = 
        List.filter (fun (f:Factor) -> f.Name = Some name |> not) factors

    /// Returns comments of a factor
    static member getComments (factor : Factor) =
        factor.Comments

    /// Applies function f on comments of a factor
    static member mapComments (f : ResizeArray<Comment> -> ResizeArray<Comment>) (factor : Factor) =
        { factor with 
            Comments = Option.map f factor.Comments}

    /// Replaces comments of a factor by given comment list
    static member setComments (factor : Factor) (comments : ResizeArray<Comment>) =
        { factor with
            Comments = Some comments }

    /// Returns factor type of a factor
    static member getFactorType (factor : Factor) =
        factor.FactorType

    /// Applies function f on factor type of a factor
    static member mapFactorType (f : OntologyAnnotation -> OntologyAnnotation) (factor : Factor) =
        { factor with 
            FactorType = Option.map f factor.FactorType}

    /// Replaces factor type of a factor by given factor type
    static member setFactorType (factor : Factor) (factorType : OntologyAnnotation) =
        { factor with
            FactorType = Some factorType }

    /// Returns the name of the factor as string if it exists
    static member tryGetName (f : Factor) =
        f.Name

    /// Returns the name of the factor as string
    static member getNameAsString (f : Factor) =
        f.NameText

    /// Returns true if the given name matches the name of the factor
    static member nameEqualsString (name : string) (f : Factor) =
        f.NameText = name

    member this.Copy() =
        let nextComments = this.Comments |> Option.map (ResizeArray.map (fun c -> c.Copy()))
        Factor.make this.Name this.FactorType nextComments

File: js/Core/Process/Factor.js
import { OntologyAnnotation_$reflection, OntologyAnnotation } from "../OntologyAnnotation.js";
import { map, defaultArg, unwrap } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { ResizeArray_map, Option_fromValueWithDefault } from "../Helper/Collections.js";
import { filter, singleton, append, exists, tryFind } from "../../fable_modules/fable-library-js.4.22.0/List.js";
import { equals } from "../../fable_modules/fable-library-js.4.22.0/Util.js";
import { Record, toString } from "../../fable_modules/fable-library-js.4.22.0/Types.js";
import { record_type, array_type, option_type, string_type } from "../../fable_modules/fable-library-js.4.22.0/Reflection.js";
import { Comment$_$reflection } from "../Comment.js";

export class Factor extends Record {
    constructor(Name, FactorType, Comments) {
        super();
        this.Name = Name;
        this.FactorType = FactorType;
        this.Comments = Comments;
    }
    static make(name, factorType, comments) {
        return new Factor(name, factorType, comments);
    }
    static create(Name, FactorType, Comments) {
        return Factor.make(Name, FactorType, Comments);
    }
    static get empty() {
        return Factor.create();
    }
    static fromString(name, term, source, accession, comments) {
        const oa = OntologyAnnotation.create(term, source, accession, unwrap(comments));
        const name_1 = Option_fromValueWithDefault("", name);
        const factorType = Option_fromValueWithDefault(new OntologyAnnotation(), oa);
        return Factor.make(name_1, factorType, undefined);
    }
    static toStringObject(factor) {
        const value = {
            TermAccessionNumber: "",
            TermName: "",
            TermSourceREF: "",
        };
        return defaultArg(map((oa) => OntologyAnnotation.toStringObject(oa), factor.FactorType), value);
    }
    get NameText() {
        const this$ = this;
        return defaultArg(this$.Name, "");
    }
    MapCategory(f) {
        const this$ = this;
        return new Factor(this$.Name, map(f, this$.FactorType), this$.Comments);
    }
    SetCategory(c) {
        const this$ = this;
        return new Factor(this$.Name, c, this$.Comments);
    }
    static tryGetByName(name, factors) {
        return tryFind((f) => equals(f.Name, name), factors);
    }
    static existsByName(name, factors) {
        return exists((f) => equals(f.Name, name), factors);
    }
    static add(factors, factor) {
        return append(factors, singleton(factor));
    }
    static removeByName(name, factors) {
        return filter((f) => !equals(f.Name, name), factors);
    }
    static getComments(factor) {
        return factor.Comments;
    }
    static mapComments(f, factor) {
        return new Factor(factor.Name, factor.FactorType, map(f, factor.Comments));
    }
    static setComments(factor, comments) {
        return new Factor(factor.Name, factor.FactorType, comments);
    }
    static getFactorType(factor) {
        return factor.FactorType;
    }
    static mapFactorType(f, factor) {
        return new Factor(factor.Name, map(f, factor.FactorType), factor.Comments);
    }
    static setFactorType(factor, factorType) {
        return new Factor(factor.Name, factorType, factor.Comments);
    }
    static tryGetName(f) {
        return f.Name;
    }
    static getNameAsString(f) {
        return f.NameText;
    }
    static nameEqualsString(name, f) {
        return f.NameText === name;
    }
    Copy() {
        const this$ = this;
        const comments = map((a) => ResizeArray_map((c) => c.Copy(), a), this$.Comments);
        return Factor.make(this$.Name, this$.FactorType, comments);
    }
    Print() {
        const this$ = this;
        return toString(this$);
    }
    PrintCompact() {
        const this$ = this;
        return "OA " + this$.NameText;
    }
}

export function Factor_$reflection() {
    return record_type("ARCtrl.Process.Factor", [], Factor, () => [["Name", option_type(string_type)], ["FactorType", option_type(OntologyAnnotation_$reflection())], ["Comments", option_type(array_type(Comment$_$reflection()))]]);
}

